include/pybind11/pytypes.h

```diff
@@ -1163,11 +1163,16 @@ class float_ : public object {
 
 class weakref : public object {
 public:
-    PYBIND11_OBJECT_DEFAULT(weakref, object, PyWeakref_Check)
+    PYBIND11_OBJECT_CVT(weakref, object, PyWeakref_Check, raw_weakref)
```

Comment:
Technically an API break. Do we care about users digging up `py::weakref::weakref()`?

Reply:
@bstaletic and I discussed this privately, btw.

I don't think `weakref()` is that useful: it's constructing a `weakref` object with a `(PyObject *) NULL` inside. pybind11 is reasonably inconsistent about which types allow this and which ones don't. I'd say it's probably a discussion for another time on the sense/nonsense of default constructor of `object`-subclasses.

I could add either `PYBIND11_OBJECT_CVT_DEFAULT` or just add a single line `weakref() = default;` here, though?

Reply:
`weakref() = default;` looks good to me, maybe with a comment
```
// Provided mainly for uniformity. Constructs a weakref object with a (PyObject *) NULL inside.
weakref() = default;
```
I almost feel the `PYBIND11_OBJECT_DEFAULT` macro isn't useful ("what default?") and getting rid of it and explicitly writing out `something() = default;` is more readable code ... but that's a question for another day, maybe.

Reply:
Agreed. This macros make things very confusing and inconsistent (some `py::object` subclasses are nullable, others aren't; some try to be safe when calling methods on them, others don't). I actually would like to doubt the validity of default constructors here in general. But maybe that's a challenge for another day.